### PR TITLE
[wmco] Start vendoring

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
The entire OpenShift org decided to
allow offline builds via vendoring.
This PR ensures that we're confirming
to that standard.

Ran `go mod vendor` to generate the vendor

### CODE REVIEW

**Comment by sebsoto** [comment]
nit: commit message mentions `PR` instead of `commit`

**Comment by ravisantoshgudimetla** [comment]
> Do we need to explicitly add this flag to the build targets for local development purposes?

Yes, I am working on it. That's why I did not cc anyone yet.

File: Makefile [hunk 1]
```diff
@@ -1,10 +1,18 @@
 all: lint build unit
 
 OUTPUT_DIR="build/_output"
+# Set the go mod vendor flags, if they're not already set
+GO ?=go
```
Comment (aravindhp) [inline Makefile:5]: Why do you need this?

File: Makefile [hunk 2]
```diff
@@ -1,10 +1,18 @@
...
+GOFLAGS ?=$(shell $(GO) env GOFLAGS)
+ifeq "$(findstring -mod=vendor,$(GOFLAGS))" "-mod=vendor"
+GO_MOD_FLAGS ?=
+else
+GO_MOD_FLAGS ?=-mod=vendor
```
Comment (aravindhp) [inline Makefile:10]: Please add a space after `?=`

File: Makefile (hunk 1 above)
Comment (ravisantoshgudimetla) [reply Makefile:5]: We use it [here](https://github.com/openshift/windows-machine-config-operator/pull/74/commits/b6ce7f7919dbc7a9032a57863ce401c5cff8c83b#diff-b67911656ef5d18c4ae36cb6741b7965R6)

File: Makefile (hunk 2 above)
Comment (ravisantoshgudimetla) [reply Makefile:10]: Let me see if that works. I am not sure...

File: Makefile (hunk 1 above)
Comment (aravindhp) [reply Makefile:5]: I understand that it is being used but why do you need it? You should be able to use `go` without it having to be a variable. Or is `$GO` present in CI and you are accounting for it in our local development environment?

File: Makefile [hunk 3]
```diff
@@ -13,11 +21,11 @@ lint:
 
 .PHONY: unit
 unit:
-	hack/unit.sh
+	hack/unit.sh ${GO_MOD_FLAGS}
 
-.PHONY: run-ci-e2e-test
+.PHONY: run-ci-e2e-test
```
Comment (aravindhp) [inline Makefile:26]: Please revert.

File: Makefile (hunk 2 above)
Comment (aravindhp) [reply Makefile:10]: You should be able to as per the [docs](https://www.gnu.org/software/make/manual/make.html#index-_003f_003d).

**Comment by sebsoto** [comment]
I'd like to see a lint job to ensure that new PRs have vendored properly. What do you think @ravisantoshgudimetla ?

File: hack/verify-vendor.sh [hunk 4]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env bash
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+WMCO_ROOT=$(pwd)
+
+# create a nice clean place to put our new vendor tree
```
Comment (aravindhp) [inline hack/verify-vendor.sh:9]: Please remove fluff like `nice clean place`. `temporary location` is sufficient. Applies everywhere.

File: hack/verify-vendor.sh [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+mkdir -p "${WMCO_ROOT}/_tmp"
+_tmpdir="$(mktemp -d "${WMCO_ROOT}/_tmp/wmco-vendor.XXXXXX")"
+
+# Copy the contents of the kube directory into the nice clean place (which is NOT shaped like a GOPATH)
```
Comment (aravindhp) [inline hack/verify-vendor.sh:13]: `shaped like a GOPATH`? What is the shape of a GOPATH? :-D

File: hack/verify-vendor.sh [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+  # Recreate the vendor tree using the nice clean set we just downloaded
+  go mod vendor
+popd > /dev/null 2>&1
+
+ret=0
+
+pushd "${WMCO_ROOT}" > /dev/null 2>&1
+  if ! _out="$(diff -Naupr -x "BUILD" -x "AUTHORS*" -x "CONTRIBUTORS*" vendor "${_wmcotmp}/vendor")"; then
+    echo "Your vendored results are different:" >&2
```
Comment (aravindhp) [inline hack/verify-vendor.sh:32]: `Vendored results...`

File: hack/verify-vendor.sh [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    echo "${_out}" >&2
+    echo "Vendor Verify failed." >&2
```
Comment (aravindhp) [inline hack/verify-vendor.sh:34]: `verify`

File: hack/verify-vendor.sh [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    echo "${_out}" > vendordiff.patch
+    echo "If you're seeing this locally, run the below command to fix your directories:" >&2
```
Comment (aravindhp) [inline hack/verify-vendor.sh:36]: Please remove `locally`

File: hack/verify-vendor.sh (hunk 5 above)
Comment (ravisantoshgudimetla) [reply hack/verify-vendor.sh:13]: like `/home/ravig/go` and having src and bin etc, if there it would be treated like a GOPATH. However I can remove it.

Comment (aravindhp) [reply hack/verify-vendor.sh]: You are talking about the directory layout. I am fine if you rephrase with that terminology.

**Comment by sebsoto** [comment]
lgtm but the lint script should probably be a different commit

**Comment by ravisantoshgudimetla** [comment]
infra flake

/retest

File: .gitignore [hunk 9]
```diff
@@ -66,6 +66,9 @@ anaconda-mode/
 
 # Dependency directories (remove the comment below to include it)
 # vendor/
+# Ignore leftovers created by `hack/` scripts
+_tmp/
+vendordiff.patch
```
Comment (aravindhp) [inline .gitignore:71]: Given you are adding this, there is also a `deploy/test/` folder that can be added here.

File: hack/verify-vendor.sh [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    echo "Vendor verify failed." >&2
+    echo "${_out}" > vendordiff.patch
+    echo "If you're seeing this, run the below command to fix your directories:" >&2
+    echo "go mod vendor" >&2
+    ret=1
+  fi
+popd > /dev/null 2>&1
+
+# clean up the temp directory
+rm -rf "${WMCO_ROOT}/_tmp"
```
Comment (aravindhp) [inline hack/verify-vendor.sh:43]: Please use `trap` and move it right after the directory is created

File: hack/verify-vendor.sh [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+WMCO_ROOT=$(pwd)
+
+# create a temporary location to put our new vendor tree
+mkdir -p "${WMCO_ROOT}/_tmp"
+_tmpdir="$(mktemp -d "${WMCO_ROOT}/_tmp/wmco-vendor.XXXXXX")"
```
Comment (aravindhp) [inline hack/verify-vendor.sh:11]: Why is `XXXXX` needed?

File: hack/verify-vendor.sh [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+tar --exclude=.git --exclude="./_*" -c . | (cd "${_wmcotmp}" && tar xf -)
+
+export GO111MODULE=on
+
+pushd "${_wmcotmp}" > /dev/null 2>&1
+  # Destroy deps in the copy of the WMCO tree
+  rm -rf ./vendor 
+
+  # Recreate the vendor tree using the clean set we just downloaded
+  go mod vendor
```
Comment (aravindhp) [inline hack/verify-vendor.sh:25]: This sort of indentation around bash builtins like `pushd` and `popd` is not required and moreover is misleading. Please remove.

File: hack/verify-vendor.sh [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    ret=1
+  fi
+popd > /dev/null 2>&1
+
+# clean up the temp directory
+rm -rf "${WMCO_ROOT}/_tmp"
+
+if [[ ${ret} -gt 0 ]]; then
+  exit ${ret}
+fi
```
Comment (aravindhp) [inline hack/verify-vendor.sh:47]: Won't `exit ${ret}` suffice without the check?

File: hack/verify-vendor.sh [hunk 14]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env bash
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+WMCO_ROOT=$(pwd)
+
+# create a temporary location to put our new vendor tree
```
Comment (aravindhp) [inline hack/verify-vendor.sh:9]: `Create`

File: hack/verify-vendor.sh [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    echo "${_out}" >&2
+    echo "Vendor verify failed." >&2
+    echo "${_out}" > vendordiff.patch
+    echo "If you're seeing this, run the below command to fix your directories:" >&2
+    echo "go mod vendor" >&2
+    ret=1
+  fi
+popd > /dev/null 2>&1
+
+# clean up the temp directory
```
Comment (aravindhp) [inline hack/verify-vendor.sh:42]: `Clean`

File: hack/verify-vendor.sh (hunk 12 above)
Comment (ravisantoshgudimetla) [reply hack/verify-vendor.sh]: Ok.

File: hack/verify-vendor.sh (hunk 11 above)
Comment (ravisantoshgudimetla) [reply hack/verify-vendor.sh:11]: We need it when templates are not given in [mktemp](https://www.gnu.org/software/autogen/mktemp.html)

File: hack/verify-vendor.sh (hunk 13 above)
Comment (ravisantoshgudimetla) [reply hack/verify-vendor.sh]: It does. Updated

File: hack/verify-vendor.sh (hunk 10 above)
Comment (ravisantoshgudimetla) [reply hack/verify-vendor.sh]: Ok.

File: .gitignore (hunk 9 above)
Comment (ravisantoshgudimetla) [reply .gitignore:71]: Done

File: Makefile [hunk 16]
```diff
@@ -1,19 +1,28 @@
 all: lint build unit
 
 OUTPUT_DIR="build/_output"
+# Set the go mod vendor flags, if they're not already set
+GOFLAGS? = $(shell go env GOFLAGS)
+ifeq "$(findstring -mod=vendor,$(GOFLAGS))" "-mod=vendor"
+GO_MOD_FLAGS ?=
+else
+GO_MOD_FLAGS ?= -mod=vendor
+endif
```
Comment (aravindhp) [inline Makefile:10]: @ravisantoshgudimetla what happens if `GOFLAGS=-mod=` before `make build` is executed? I am observing locally that it remains as is and does not become `-mod=vendor`.

**Pull request merged**